pennylane/spin/lattice.py

```diff
@@ -0,0 +1,196 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions and classes to create a
+:class:`~pennylane.spin.lattice` object. This object stores all
+the necessary information about a lattice.
+"""
+import itertools
+
+import numpy as np
+from scipy.spatial import cKDTree
+
+# pylint: disable=too-many-arguments, too-many-instance-attributes
+# pylint: disable=use-a-generator, too-few-public-methods
+
+
+class Lattice:
+    r"""Constructs a Lattice object.
+
+    Args:
+       L: Number of unit cells in a direction, it is a list depending on the dimensions of the lattice.
+       unit_cell: Primitive vectors for the lattice.
+       basis: Initial positions of spins.
+       boundary_condition: defines boundary conditions, boolean or series of bools with dimensions same as L.
+       neighbour_order: Range of neighbours a spin interacts with.
+       distance_tol: Distance below which spatial points are considered equal for the purpose of identifying nearest neighbours.
+
+    Returns:
+       Lattice object
+    """
+
+    def __init__(
+        self,
+        L,
+        unit_cell,
+        basis=None,
+        boundary_condition=False,
+        neighbour_order=1,
+        distance_tol=1e-5,
+    ):
+
+        self.L = np.asarray(L)
+        self.n_dim = len(L)
+        self.unit_cell = np.asarray(unit_cell)
+        if basis is None:
+            basis = np.zeros(self.unit_cell.shape[0])[None, :]
+
+        self.basis = np.asarray(basis)
+        self.n_sl = len(self.basis)
+        self.n_sites = np.prod(L) * self.n_sl
+
+        if isinstance(boundary_condition, bool):
+            boundary_condition = [boundary_condition for _ in range(self.n_dim)]
+
+        self.boundary_condition = boundary_condition
+        self._test_input_accuracy()
```

Comment:
Even if you would want to keep it as a separate function (to maintain code readability), ensure it takes in the given `inputs` and is placed on top. This will allow for a preferred fail-fast-and-first approach for the invalid inputs.